### ISSUE TITLE
Script/Automate Performance Benchmarks

### ISSUE THREAD

**Issue by errm** [traefik/traefik#2694]
It would be good to have a script to do some sort of performance benchmarking...

Ideally against some known quantity (e.g. nginx and/or haproxy).

This can start off as a script that can be run manually in order to make it a reproducible task to update numbers on website / docs / blogposts etc.

It would also be nice to run some performance benchmarks as part of CI to try and track performance regressions, hopefully the same thing could be used for each.

See: #2690

**Comment by timoreimann** [comment]
Experience taught me that benchmarks run outside of dedicated machines (especially on local machines) are harder to get right and difficult to compare. I'd opt for having dedicated machines well optimized for the task at hand to guarantee reliable results and reproducibility.

**Comment by errm** [comment]
Agreed...

I am sure if we tried running benchmarks on travis/circle there will be a lot of variability ...

It still makes sense to get something broadly reproducible (i.e. scripted) that we can run... If we need to add some infrastructure to support getting consistent results then so be it...

I find having a performance script (of whatever sophistication) + git bisect makes tracking down performance regressions much simpler ...

**Comment by timoreimann** [comment]
Yep, I'm completely with you on those points. What I basically wanted to point out is that an isolated environment for performance testing is needed so that we can objectively compare different versions / implementations of Traefik. User-provided results should be possible but be taken with a grain of salt when comparing across different source machines.

AFAIK, Travis and other CI providers support running on dedicated machines to prevent interfering with other processes. And if not, we can still hook up some AWS or GCE machines. Cost can be minimized by running only at certain times if we need to.

Tool-wise, I like `hey` a lot for benchmarking maximum QPS and percentiles.

**Comment by errm** [comment]
`hey` looks like a good option to get something put together quickly...

**Comment by 27Bslash6** [comment]
Please forgive a shameless plug, but we're using a docker Locust setup using on-demand GKE clusters in our CI and find it to be a very reliable load test tool. See https://github.com/greenpeace/planet4-locust

Shouldn't take much time to spin up and happy to help if required.

( Found this thread while debugging high CPU usage under load #2673 )

**Comment by amq** [comment]
See also https://github.com/jwilder/nginx-proxy/issues/520#issuecomment-338472441 (and my next comment there).

**Comment by vairakkumar-vairaperumal** [comment]
https://www.loggly.com/blog/benchmarking-5-popular-load-balancers-nginx-haproxy-envoy-traefik-and-alb/

**Comment by rtribotte** [comment]
Hello there,

This proposal has been open for a while with minimal traction. We won't be adding it to our roadmap at this time and prefer to close it for now.

If circumstances change or interest grows, we can always reopen this later.
